File: 901-online-stock-span/901-online-stock-span.py
class StockSpanner:

    def __init__(self):
        self.spans = []
        # self.spans[n] = (price, span, start)

    def next(self, price: int) -> int:
        span = 0
        i = len(self.spans) - 1
        
        if i >= 0:
            curr = self.spans[i]
        
        while (i >= 0 and curr[0] <= price):
            span += curr[1]
            i = curr[2]
            curr = self.spans[i]
        
        span += 1
        
        self.spans.append((price, span, i))
        
        return span


# Your StockSpanner object will be instantiated and called as such:
# obj = StockSpanner()
# param_1 = obj.next(price)

File: 901-online-stock-span/901-online-stock-span.cpp
class StockSpanner {
public:
    
    vector<pair<int,int>> spans;
    
    StockSpanner() {
        spans = vector<pair<int,int>>();
    }
    
    int next(int price) {
        
        int curr_span = 0;
        int i = spans.size() - 1;
        
        while (i >= 0 && spans[i].first <= price){
            curr_span += spans[i].second;
            i -= spans[i].second;
        }
        
        curr_span++;
        
        spans.push_back(pair<int,int>(price, curr_span));
        
        return curr_span;
        
    }
};

/**
 * Your StockSpanner object will be instantiated and called as such:
 * StockSpanner* obj = new StockSpanner();
 * int param_1 = obj->next(price);
 */